File: adventofcode/2021/day-6/part1.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include <cstdint>

using namespace std;


void SplitString(string s, vector<uint_fast8_t> &v){
	
	string temp = "";
	for(int i=0;i<s.length();++i){
		
		if(s[i]==','){
			v.push_back(std::stoi(temp));
			temp = "";
		}
		else{
			temp.push_back(s[i]);
		}
		
	}
	/* v.push_back(temp); */
	v.push_back(std::stoi(temp));
}


void PrintVector(vector<uint_fast8_t> v)
{
	for(int i=0;i<v.size();++i)
		cout<<v[i];
	cout<<"\n";
}

int main(int argc, char *argv[]) 
{
    std::ifstream fdata("input.txt");
    /* std::ifstream fdata("input1.txt"); */
    vector<uint_fast8_t> lanfish_timers;

    std::string fishdata;
    std::getline(fdata, fishdata);

    SplitString(fishdata, lanfish_timers);
    PrintVector(lanfish_timers);

    int days_of_simulation = 256;  // part1: 80, part2: 256 (exponential growth!)
    for(int i = 0; i < days_of_simulation; i++)
    {
        cout << "calculation day: " << i+1 << endl;
        int current_len = lanfish_timers.size();
        for(int fish_i = 0; fish_i < current_len; fish_i++)
        {
            bool newf = false;
            if(lanfish_timers[fish_i] == 0)
            {
                lanfish_timers[fish_i] = 6;
                lanfish_timers.push_back(8);
                newf = true;
            }

            if(newf == false)
            {
                lanfish_timers[fish_i] -= 1;
            }
        }
        
        cout << "After " << i+1 << " days: " << lanfish_timers.size() << endl;
    }
    fdata.close(); 
}

File: adventofcode/2021/day-6-Lanternfish/part1.py
#! /usr/bin/env python3

fdata = open("input.txt", 'r')
# fdata = open("input1.txt", 'r')

lanfish_timers = [int(i) for i in fdata.readline().rstrip().split(',')]
print(lanfish_timers)

# Run simulation for X days
days_of_simulation = 256  # part1: 80, part2: 256 (exponential growth!)
for i in range(0, days_of_simulation):
    print("calculating day:",i+1)
    # calculate change for 1 day for each fish
    current_len = len(lanfish_timers)
    # for idx in range(0, len(lanfish_timers)):
    for idx in range(0, current_len):
        new = False
        if lanfish_timers[idx] == 0:
            lanfish_timers[idx] = 6
            lanfish_timers.append(8)
            new = True

        if new is False:
            lanfish_timers[idx] -= 1

    # print(f"After {i+1} days: ", lanfish_timers)
    print(f"After {i+1} days: ", len(lanfish_timers))

final_number_of_fish = len(lanfish_timers)
print(f"number of fish after {days_of_simulation} days is {final_number_of_fish}")

fdata.close()
